rfcs/0004-insecure-http/README.md

```diff
@@ -0,0 +1,173 @@
+          - --storage-path=/data
+          - --storage-adv-addr=source-controller.$(RUNTIME_NAMESPACE).svc.cluster.local.
+          - --allow-insecure-http=false
+```
+
+### Objects
+Some Flux objects, like `GitRepository`, provide a field for specifying a URL, and the URL would contain the scheme.
+In such cases, the scheme can be used for inferring the transport type of the connection and consequently,
+whether to use HTTP or HTTPS connections for that object.
+But there are a few objects that don't allow such behavior, for example:
+
+* `ImageRepository`: It provides a field, `.spec.image`, which is used for specifying the URL of the image present on
+a container registry. But any URL containing a scheme is considered invalid and HTTPS is the default transport used.
+This prevents users from using images present on insecure registries.
+* OCI `HelmRepository`: When using an OCI registry as a Helm repository, the `.spec.url` is expected to begin with `oci://`.
+Since the scheme part of the URL is used to specify the type of `HelmRepository`, there is no way for users to specify
+that the registry is hosted at a non-TLS endpoint.
+
+For such objects, we shall introduce a new boolean field `.spec.insecure`, which shall be `false` by default. Users that
+need their object to point to an HTTP endpoint, can set this to `true`.
+
+### CLI
+The Flux CLI offers several commands for creating Flux specific resources. Some of these commands may involve specifying
+an endpoint such as creating an `OCIRepository`:
+
+```sh
+ flux create source oci podinfo \
+    --url=oci://ghcr.io/stefanprodan/manifests/podinfo \
+    --tag=6.1.6 \
+    --interval=10m
+```
+
+Since these commands essentially create object definitions, the CLI should offer a boolean flag `--insecure`
+for relevant objects, which will be used for specifying the value of `.spec.insecure` of such objects.
+
+### Precedence & Validitiy
+Objects with `.spec.insecure` as `true `will only be allowed if HTTP connections are allowed at the controller
+level.
+Similarly, an object can have`.spec.insecure` as `true` only if the Saas/Cloud provider allows HTTP connections.
+For example using a `Bucket` with its `.spec.provider` set to `azure` would be invalid, since Azure doesn't allow
+HTTP connections.
+
+
+### User Stories
+
+#### Story 1
+> As a cluster admin of a multi-tenant cluster, I want to ensure all controllers access endpoints using only HTTPS
+> regardless of tenants' object definitions.
+
+Apply a `kustomize` patch which prevents the use of HTTP connections:
+
+```yaml
+apiVersion: kustomize.config.k8s.io/v1beta1
+kind: Kustomization
+resources:
+  - gotk-components.yaml
+  - gotk-sync.yaml
+patches:
+  - patch: |
+      - op: add
+        path: /spec/template/spec/containers/0/args/-
+        value: --allow-insecure-http=false
+    target:
+      kind: Deployment
+      name: "(kustomize-controller|helm-controller|source-controller|notification-controller)"
```

Comment:
Agree with you on helm-controller.
Regarding kustomize-controller, the default behavior is to allow remote bases, so admins might want to make sure that only HTTPS is used to access those remote bases.

Reply:
We'll need to fork kustomize to add this filter, instead I propose we recommend users to disable remote bases all together as they are insecure by default HTTPS or nor.

Reply:
I think this would still be useful in Kustomize Controller to enforce remote applies. In this case, the controller will error if `insecure-skip-tls-verify` inside the `.spec.kubeconfig` is not aligned with the controller level flag.

Reply:
They are orthogonal. My point was not related to remote bases, it was related to **remote applies**. The flag is a controller level way to enforce that all outward communications are using TLS, which in this case would be considered when applying reconciliations against remote clusters.

Reply:
Both kustomize and helm controllers have the flag `--insecure-kubeconfig-tls` witch by default is set to `false` to prevent insecure connections to remote clusters. Are you proposing to drop this flag and replace it with a new one which would allow insecure connections by default?

Reply:
`--inscecure-kubeconfig-tls` would skip TLS verification, but the protocol used would still HTTPS. I don't think adding `--allow-insecure-http` makes sense for remote applies as kube-apiserver by default is served over HTTPS and while technically it can be served over plain HTTP, the option for that configuration was disabled quite a while back: https://github.com/kubernetes/kubernetes/pull/65830